platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/UnsignedEvent.java

```diff
@@ -161,53 +181,112 @@ public BaseEventHashedData(
         // it is possible to skip transaction array and move on to the next object
-        out.writeInt(getSerializedLength(transactions, true, false));
+        out.writeInt(PayloadUtils.getLegacyObjectSize(payloads));
         // transactions may include both system transactions and application transactions
         // so writeClassId set to true and allSameClass set to false
-        out.writeSerializableArray(transactions, true, false);
-    }
-
-    @Override
-    public void deserialize(final SerializableDataInputStream in, final int version) throws IOException {
-        final TransactionConfig transactionConfig = ConfigurationHolder.getConfigData(TransactionConfig.class);
-        deserialize(in, version, transactionConfig.maxTransactionCountPerEvent());
+        final boolean allSameClass = false;
+        out.writeInt(payloads.size());
+        if (!payloads.isEmpty()) {
+            out.writeBoolean(allSameClass);
+        }
+        for (final EventPayload payload : payloads) {
+            switch (payload.payload().kind()) {
+                case APPLICATION_PAYLOAD:
+                    out.writeLong(APPLICATION_TRANSACTION_CLASS_ID);
+                    out.writeInt(APPLICATION_TRANSACTION_VERSION);
+                    final Bytes bytes = payload.payload().as();
+                    out.writeInt((int) bytes.length());
+                    bytes.writeTo(out);
+                    break;
+                case STATE_SIGNATURE_PAYLOAD:
+                    final StateSignaturePayload stateSignaturePayload =
+                            payload.payload().as();
+
+                    out.writeLong(STATE_SIGNATURE_CLASS_ID);
+                    out.writeInt(STATE_SIGNATURE_VERSION);
+                    out.writeInt((int) stateSignaturePayload.signature().length());
+                    stateSignaturePayload.signature().writeTo(out);
+
+                    out.writeInt((int) stateSignaturePayload.hash().length());
+                    stateSignaturePayload.hash().writeTo(out);
+
+                    out.writeLong(stateSignaturePayload.round());
+                    out.writeInt(Integer.MIN_VALUE); // epochHash is always null
+                    break;
+                default:
+                    throw new IOException(
+                            "Unknown payload type: " + payload.payload().kind());
+            }
+        }
     }
 
-    private void deserialize(
-            @NonNull final SerializableDataInputStream in, final int version, final int maxTransactionCount)
+    /**
+     * Deserialize the event.
+     *
+     * @param in the stream from which this object is to be read
+     * @return the deserialized event
+     * @throws IOException if unsupported payload types are encountered
+     */
+    public static UnsignedEvent deserialize(final SerializableDataInputStream in, final int version)
             throws IOException {
+        in.readInt(); // read and ignore version
```

Comment:
i think we should throw if it doesnt match the expected version

Reply:
done